### ISSUE TITLE
Shallow clone ownership

### ISSUE THREAD

**Issue by yth** [MHebes/cs4500-a1-part2#33]
If map2 is a shallow copy of map1, who has ultimate ownership of the keys and values since they need to point to the same objects per last comment by gyroknight on #4? In other words, which object is responsible for freeing them?

**Comment by gyroknight** [comment]
The original map is responsible for the values. The shallow copy can be seen as a viewing window into the original map; you can safely look inside and see all the key and values without any effect, but if you decide to alter any of the keys or values, the window gets broken and glass gets everywhere.

**Comment by yth** [comment]
If that's the case, what's preventing the copy from calling its destructor when it goes away first? What's prevent the original from being destroyed, and have the copy access memory it's not entitled to?

I feel I don't understand your design and the motivation behind this decision.

**Comment by gyroknight** [comment]
You're correct; I actually just ran into this issue while working on our implementation of Map. The user should ultimately hold ownership of the objects, deleting the map does not affect its values.

**Comment by yth** [comment]
That seems a little reckless since we are all fallible. I hope you won't mind I change your design a little to be a little more defensive.

It might be too late to change the official design. It took me a while to think through the implications, and I might still be missing something. So, I really wasn't off much help. Still, I wanted to let you know my direction of thinking.

**Comment by gyroknight** [comment]
It's certainly possible to implement the original design after a bit more thought. If you want to make the copy not destroy the objects, you could have a field (maybe `isCopy` or something) that changes the behavior of the destructor. In fact, after a bit more thought, I think this is the behavior we want to target.

**Comment by yth** [comment]
I think you want a copy-on-write system.

That might require a lot more specifications. E.g. should the copy see changes the original makes after the copy is made, when does the break of the mirroring happens, what are all the conditions that triggers this, and how to inform both the original and the copy that the break is happening or has just happened?
